File: web/src/lib/game/wordle/wordle.ts
// Client-side Wordle game — optimistic preview implementation.
// Mirrors server/internal/game/wordle/wordle.go.
// SERVER IS AUTHORITATIVE: this is used only for instant UI feedback before
// the server response arrives. The server reply always overrides local state.

import type { Game, Move, Result, State } from '../game';
import { score, WORD_LEN, type Color } from './colors';

export const MAX_ATTEMPTS = 6;

// ── Types ─────────────────────────────────────────────────────────────────────

export interface WordleMove extends Move {
	readonly guess: string;
}

export interface WordleState extends State {
	readonly guesses: readonly string[];
	readonly hints: readonly (readonly Color[])[];
	readonly attemptsRemaining: number;
	readonly won: boolean;
	readonly lost: boolean;
	// solution only populated when isTerminal (server-authoritative reveal).
	readonly solution: string;
}

// ── Initial state ─────────────────────────────────────────────────────────────

/** createInitialState returns a fresh game state for optimistic preview. */
export function createInitialState(): WordleState {
	return {
		guesses: [],
		hints: [],
		attemptsRemaining: MAX_ATTEMPTS,
		won: false,
		lost: false,
		solution: '',
		get isTerminal(): boolean {
			return this.won || this.lost;
		}
	};
}

// ── Game implementation ───────────────────────────────────────────────────────

/** WordleGame implements the pluggable Game interface for client-side preview. */
export class WordleGame implements Game<WordleState, WordleMove> {
	private readonly dictionary: ReadonlySet<string>;

	constructor(dictionary: readonly string[]) {
		this.dictionary = new Set(dictionary.map((w) => w.toUpperCase()));
	}

	validate(state: WordleState, move: WordleMove): string | null {
		if (state.isTerminal) return 'Game is already over';
		const upper = move.guess.toUpperCase();
		if (upper.length !== WORD_LEN) return `Guess must be exactly ${WORD_LEN} letters`;
		if (!this.dictionary.has(upper)) return 'Not in word list';
		return null;
	}

	apply(state: WordleState, move: WordleMove): WordleState {
		const upper = move.guess.toUpperCase();
		// solution is '' for optimistic preview (unknown client-side); use empty
		// string so hints are all gray — server reply will correct them.
		const hint = state.solution ? score(upper, state.solution) : new Array<Color>(WORD_LEN).fill('gray');
		const newGuesses = [...state.guesses, upper];
		const newHints = [...state.hints, hint];
		const remaining = state.attemptsRemaining - 1;
		const won = state.solution !== '' && upper === state.solution;
		const lost = !won && remaining === 0;

		return {
			guesses: newGuesses,
			hints: newHints,
			attemptsRemaining: remaining,
			won,
			lost,
			solution: state.solution,
			get isTerminal(): boolean {
				return this.won || this.lost;
			}
		};
	}

	result(state: WordleState): Result {
		return {
			won: state.won,
			attemptsUsed: state.guesses.length
		};
	}
}

File: server/internal/game/wordle/wordle.go
package wordle

import (
	"errors"
	"strings"

	"github.com/tiennm99/dleague/shared/game"
	dleaguev1 "github.com/tiennm99/dleague/shared/pb/dleague/v1"
)

const (
	// MaxAttempts is the maximum number of guesses a player may submit.
	MaxAttempts = 6
	// WordLen is the required word length for guesses and solutions.
	WordLen = 5
)

// Sentinel errors for validation — exported so handlers can check with errors.Is.
var (
	ErrWrongLength    = errors.New("wordle: guess must be exactly 5 letters")
	ErrNotInDictionary = errors.New("wordle: guess not in word list")
	ErrGameOver       = errors.New("wordle: game is already over")
)

// Wordle holds the mutable state of one server-authoritative Wordle session.
// solution is never sent to the client until the game is terminal.
type Wordle struct {
	solution          string   // upper-case 5-letter target word
	guesses           []string // submitted guesses in order
	hints             [][]Color
	attemptsRemaining int
	won               bool
	lost              bool
}

// New creates a fresh Wordle session for the given solution.
// solution must be exactly WordLen upper-case ASCII letters.
func New(solution string) *Wordle {
	return &Wordle{
		solution:          strings.ToUpper(solution),
		attemptsRemaining: MaxAttempts,
	}
}

// Validate checks whether the guess is acceptable without mutating state.
// Both length and dictionary checks are performed before returning any error
// to avoid leaking timing about which check failed.
func (w *Wordle) Validate(guess string, dict []string) error {
	// Normalise to upper-case before length check — consistent with Apply.
	upper := strings.ToUpper(guess)

	// Length and dictionary checks both evaluate before returning so callers
	// cannot distinguish which failed by measuring response time.
	lengthOK := len(upper) == WordLen
	inDict := contains(dict, upper)

	if !lengthOK {
		return ErrWrongLength
	}
	if !inDict {
		return ErrNotInDictionary
	}
	return nil
}

// Apply records the guess, scores it, and advances game state.
// Call Validate before Apply; Apply does not re-validate.
func (w *Wordle) Apply(guess string) error {
	if w.IsTerminal() {
		return ErrGameOver
	}
	upper := strings.ToUpper(guess)
	colors := Score(upper, w.solution)
	w.guesses = append(w.guesses, upper)
	w.hints = append(w.hints, colors)
	w.attemptsRemaining--

	if upper == w.solution {
		w.won = true
	} else if w.attemptsRemaining == 0 {
		w.lost = true
	}
	return nil
}

// IsTerminal reports whether the game has reached a terminal state.
func (w *Wordle) IsTerminal() bool {
	return w.won || w.lost
}

// Result returns a game.Result; only valid when IsTerminal() is true.
func (w *Wordle) Result() game.Result {
	return game.Result{
		Won:          w.won,
		AttemptsUsed: len(w.guesses),
	}
}

// ToProto converts the current state to a wire-safe WordleState proto.
// solution is only populated when the game is terminal.
func (w *Wordle) ToProto() *dleaguev1.WordleState {
	hints := make([]*dleaguev1.WordleHint, len(w.hints))
	for i, h := range w.hints {
		colors := make([]dleaguev1.Color, len(h))
		copy(colors, h)
		hints[i] = &dleaguev1.WordleHint{Colors: colors}
	}
	state := &dleaguev1.WordleState{
		Guesses:           append([]string(nil), w.guesses...),
		Hints:             hints,
		AttemptsRemaining: int32(w.attemptsRemaining), //nolint:gosec
		Won:               w.won,
		Lost:              w.lost,
	}
	// Reveal solution only on terminal state to prevent client-side cheating.
	if w.IsTerminal() {
		state.Solution = w.solution
	}
	return state
}

// contains is a simple O(n) linear search. Dictionaries are small enough
// (<5000 words) that a map is unnecessary for MVP; Phase 10 can optimise.
func contains(dict []string, word string) bool {
	for _, w := range dict {
		if w == word {
			return true
		}
	}
	return false
}
